utils/merge.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package utils
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Merge will return a new struct/map/slice of the same type as base and patch, with patch merged into base.
+// Specifically, patch's values will be preferred except when patch's value is `nil`.
+// Restrictions/guarantees:
+//   - base and patch will not be modified
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - if slices are different, they will be merged according to the following rules:
+//       - start with elements from base (if any)
+//		 - then for each item, patch will overwrite if different
+//		 - and then add the extra items from patch (if any)
+//   - if maps are different, they will be merged according to the following rules:
+//       - keys in base are added first
+//       - keys in patch are added after, overwriting existing keys with values from patch
+//       - reference values (eg. slice/ptr/map/chan) will not be cloned
+//   - channel values are not supported at the moment
+//
+// Usage: callers need to cast the returned interface back into the original type, eg:
+// func mergeTestStruct(base, patch *testStruct) (*testStruct, error) {
+//     ret, err := merge(base, patch)
+//     if err != nil {
+//         return nil, err
+//     }
+//     retTS := ret.(testStruct)
+//     return &retTS, nil
+// }
+func Merge(base interface{}, patch interface{}) (interface{}, error) {
+	if reflect.TypeOf(base) != reflect.TypeOf(patch) {
+		return nil, fmt.Errorf("cannot merge different types. base type: %s, patch type: %s",
+			reflect.TypeOf(base), reflect.TypeOf(patch))
+	}
+
+	baseType := reflect.TypeOf(base)
+	baseVal := reflect.ValueOf(base)
+	patchVal := reflect.ValueOf(patch)
+	if baseType.Kind() == reflect.Ptr {
+		baseType = baseType.Elem()
+		baseVal = baseVal.Elem()
+		patchVal = patchVal.Elem()
+	}
+
+	ret := reflect.New(baseType)
+
+	val, ok := merge(baseVal, patchVal)
+	if ok {
+		ret.Elem().Set(val)
+	}
+	return ret.Elem().Interface(), nil
+}
+
+// merge recursively merges patch into base and returns the new struct, ptr, slice/map, or value
+func merge(base, patch reflect.Value) (reflect.Value, bool) {
```

Comment:
I'm wondering if we can eliminate the `bool` return parameter here: it seems we only use it to signal that a `nil` value was returned, but the caller had already initialized to that value anyway, so there's no loss to just assigning it again. Thoughts?

Reply:
I tried and was banging my head on the wall about this. `reflect.Value` can't be nil, so you always have to return something, and the zero value of a concrete type can't be nil (of course). So I couldn't figure out how to set the pointer to nil without that extra `bool` return param.

This isn't a problem when the value is a pointer (`merge.go` line 108-109) since we check for nil value above and return the newly created (and thus nil) mergedPtr if we're setting to nil (line 100).